// @shopify/shopify-api/lib/logger/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
var tslib_1 = require("tslib");
var semver_1 = tslib_1.__importDefault(require("semver"));
var types_1 = require("../types");
var error_1 = require("../error");
var version_1 = require("../version");
var log_1 = require("./log");
function logger(config) {
    var _this = this;
    var logFunction = (0, log_1.log)(config);
    return {
        log: logFunction,
        debug: function (message, context) {
            if (context === void 0) { context = {}; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, logFunction(types_1.LogSeverity.Debug, message, context)];
            }); });
        },
        info: function (message, context) {
            if (context === void 0) { context = {}; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, logFunction(types_1.LogSeverity.Info, message, context)];
            }); });
        },
        warning: function (message, context) {
            if (context === void 0) { context = {}; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, logFunction(types_1.LogSeverity.Warning, message, context)];
            }); });
        },
        error: function (message, context) {
            if (context === void 0) { context = {}; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, logFunction(types_1.LogSeverity.Error, message, context)];
            }); });
        },
        deprecated: deprecated(logFunction),
    };
}
exports.logger = logger;
function deprecated(logFunction) {
    return function (version, message) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (semver_1.default.gte(version_1.SHOPIFY_API_LIBRARY_VERSION, version)) {
                    throw new error_1.FeatureDeprecatedError("Feature was deprecated in version ".concat(version));
                }
                return [2 /*return*/, logFunction(types_1.LogSeverity.Warning, "[Deprecated | ".concat(version, "] ").concat(message))];
            });
        });
    };
}
