// @shopify/shopify-api/rest/admin/2023-01/index.js
"use strict";
/***********************************************************************************************************************
* This file is auto-generated. If you have an issue, please create a GitHub issue.                                     *
***********************************************************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.restResources = void 0;
var abandoned_checkout_1 = require("./abandoned_checkout");
var access_scope_1 = require("./access_scope");
var apple_pay_certificate_1 = require("./apple_pay_certificate");
var application_charge_1 = require("./application_charge");
var application_credit_1 = require("./application_credit");
var article_1 = require("./article");
var asset_1 = require("./asset");
var assigned_fulfillment_order_1 = require("./assigned_fulfillment_order");
var balance_1 = require("./balance");
var blog_1 = require("./blog");
var cancellation_request_1 = require("./cancellation_request");
var carrier_service_1 = require("./carrier_service");
var checkout_1 = require("./checkout");
var collect_1 = require("./collect");
var collection_1 = require("./collection");
var collection_listing_1 = require("./collection_listing");
var comment_1 = require("./comment");
var country_1 = require("./country");
var currency_1 = require("./currency");
var custom_collection_1 = require("./custom_collection");
var customer_1 = require("./customer");
var customer_address_1 = require("./customer_address");
var customer_saved_search_1 = require("./customer_saved_search");
var deprecated_api_call_1 = require("./deprecated_api_call");
var discount_code_1 = require("./discount_code");
var dispute_1 = require("./dispute");
var draft_order_1 = require("./draft_order");
var event_1 = require("./event");
var fulfillment_1 = require("./fulfillment");
var fulfillment_event_1 = require("./fulfillment_event");
var fulfillment_order_1 = require("./fulfillment_order");
var fulfillment_request_1 = require("./fulfillment_request");
var fulfillment_service_1 = require("./fulfillment_service");
var gift_card_1 = require("./gift_card");
var gift_card_adjustment_1 = require("./gift_card_adjustment");
var image_1 = require("./image");
var inventory_item_1 = require("./inventory_item");
var inventory_level_1 = require("./inventory_level");
var location_1 = require("./location");
var locations_for_move_1 = require("./locations_for_move");
var marketing_event_1 = require("./marketing_event");
var metafield_1 = require("./metafield");
var mobile_platform_application_1 = require("./mobile_platform_application");
var order_1 = require("./order");
var order_risk_1 = require("./order_risk");
var page_1 = require("./page");
var payment_1 = require("./payment");
var payment_gateway_1 = require("./payment_gateway");
var payment_transaction_1 = require("./payment_transaction");
var payout_1 = require("./payout");
var policy_1 = require("./policy");
var price_rule_1 = require("./price_rule");
var product_1 = require("./product");
var product_listing_1 = require("./product_listing");
var product_resource_feedback_1 = require("./product_resource_feedback");
var province_1 = require("./province");
var recurring_application_charge_1 = require("./recurring_application_charge");
var redirect_1 = require("./redirect");
var refund_1 = require("./refund");
var report_1 = require("./report");
var resource_feedback_1 = require("./resource_feedback");
var script_tag_1 = require("./script_tag");
var shipping_zone_1 = require("./shipping_zone");
var shop_1 = require("./shop");
var smart_collection_1 = require("./smart_collection");
var storefront_access_token_1 = require("./storefront_access_token");
var tender_transaction_1 = require("./tender_transaction");
var theme_1 = require("./theme");
var transaction_1 = require("./transaction");
var usage_charge_1 = require("./usage_charge");
var user_1 = require("./user");
var variant_1 = require("./variant");
var webhook_1 = require("./webhook");
exports.restResources = {
    AbandonedCheckout: abandoned_checkout_1.AbandonedCheckout,
    AccessScope: access_scope_1.AccessScope,
    ApplePayCertificate: apple_pay_certificate_1.ApplePayCertificate,
    ApplicationCharge: application_charge_1.ApplicationCharge,
    ApplicationCredit: application_credit_1.ApplicationCredit,
    Article: article_1.Article,
    Asset: asset_1.Asset,
    AssignedFulfillmentOrder: assigned_fulfillment_order_1.AssignedFulfillmentOrder,
    Balance: balance_1.Balance,
    Blog: blog_1.Blog,
    CancellationRequest: cancellation_request_1.CancellationRequest,
    CarrierService: carrier_service_1.CarrierService,
    Checkout: checkout_1.Checkout,
    Collect: collect_1.Collect,
    Collection: collection_1.Collection,
    CollectionListing: collection_listing_1.CollectionListing,
    Comment: comment_1.Comment,
    Country: country_1.Country,
    Currency: currency_1.Currency,
    CustomCollection: custom_collection_1.CustomCollection,
    Customer: customer_1.Customer,
    CustomerAddress: customer_address_1.CustomerAddress,
    CustomerSavedSearch: customer_saved_search_1.CustomerSavedSearch,
    DeprecatedApiCall: deprecated_api_call_1.DeprecatedApiCall,
    DiscountCode: discount_code_1.DiscountCode,
    Dispute: dispute_1.Dispute,
    DraftOrder: draft_order_1.DraftOrder,
    Event: event_1.Event,
    Fulfillment: fulfillment_1.Fulfillment,
    FulfillmentEvent: fulfillment_event_1.FulfillmentEvent,
    FulfillmentOrder: fulfillment_order_1.FulfillmentOrder,
    FulfillmentRequest: fulfillment_request_1.FulfillmentRequest,
    FulfillmentService: fulfillment_service_1.FulfillmentService,
    GiftCard: gift_card_1.GiftCard,
    GiftCardAdjustment: gift_card_adjustment_1.GiftCardAdjustment,
    Image: image_1.Image,
    InventoryItem: inventory_item_1.InventoryItem,
    InventoryLevel: inventory_level_1.InventoryLevel,
    Location: location_1.Location,
    LocationsForMove: locations_for_move_1.LocationsForMove,
    MarketingEvent: marketing_event_1.MarketingEvent,
    Metafield: metafield_1.Metafield,
    MobilePlatformApplication: mobile_platform_application_1.MobilePlatformApplication,
    Order: order_1.Order,
    OrderRisk: order_risk_1.OrderRisk,
    Page: page_1.Page,
    Payment: payment_1.Payment,
    PaymentGateway: payment_gateway_1.PaymentGateway,
    PaymentTransaction: payment_transaction_1.PaymentTransaction,
    Payout: payout_1.Payout,
    Policy: policy_1.Policy,
    PriceRule: price_rule_1.PriceRule,
    Product: product_1.Product,
    ProductListing: product_listing_1.ProductListing,
    ProductResourceFeedback: product_resource_feedback_1.ProductResourceFeedback,
    Province: province_1.Province,
    RecurringApplicationCharge: recurring_application_charge_1.RecurringApplicationCharge,
    Redirect: redirect_1.Redirect,
    Refund: refund_1.Refund,
    Report: report_1.Report,
    ResourceFeedback: resource_feedback_1.ResourceFeedback,
    ScriptTag: script_tag_1.ScriptTag,
    ShippingZone: shipping_zone_1.ShippingZone,
    Shop: shop_1.Shop,
    SmartCollection: smart_collection_1.SmartCollection,
    StorefrontAccessToken: storefront_access_token_1.StorefrontAccessToken,
    TenderTransaction: tender_transaction_1.TenderTransaction,
    Theme: theme_1.Theme,
    Transaction: transaction_1.Transaction,
    UsageCharge: usage_charge_1.UsageCharge,
    User: user_1.User,
    Variant: variant_1.Variant,
    Webhook: webhook_1.Webhook,
};
